### PR TITLE
[azurerm_storage_account] support for the `enable_advanced_threat_protection` property

### LINKED ISSUE #3670: New-Feature-Request Storage Threat Detection Policy
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description
Enable threat detection on storage accounts.

### New or Affected Resource(s)
* azurerm_storage_account

### Potential Terraform Configuration

Copy the way it has been implemented for [`azurerm_sql_database`](https://www.terraform.io/docs/providers/azurerm/r/sql_database.html#threat_detection_policy)

```hcl
threat_detection_policy = {}
```

### References
Go SDK does not appear to support it.

Azure API does have it:
https://docs.microsoft.com/en-us/rest/api/securitycenter/advancedthreatprotection/create

### PR DESCRIPTION
**Pull request by benjamin37**
I'm not sure about the following point of this pull request. Advanced Threat Protection is a feature you can set on a storage account. However it is set with a separate client `AdvancedThreatProtectionClient` and not with the `StorageServiceClient`. Is it common practice to create a new resource for this setting instead of adding it to the storage account itself?

Relates to issue https://github.com/terraform-providers/terraform-provider-azurerm/issues/3670

(fixes #3670)

### CODE REVIEW

File: azurerm/resource_arm_storage_account.go [hunk 1]
```diff
@@ -548,6 +556,19 @@ func resourceArmStorageAccountCreate(d *schema.ResourceData, meta interface{}) e
 	log.Printf("[INFO] storage account %q ID: %q", storageAccountName, *account.ID)
 	d.SetId(*account.ID)
 
+	enableAdvanceThreatProtection := d.Get("enable_advanced_threat_protection").(bool)
+
+	advancedThreatProtectionSetting := security.AdvancedThreatProtectionSetting{
+		AdvancedThreatProtectionProperties: &security.AdvancedThreatProtectionProperties{
+			IsEnabled: &enableAdvanceThreatProtection,
```
Comment (katbyte) [inline azurerm/resource_arm_storage_account.go]: _minor_ you could do:
```suggestion
			IsEnabled: utils.bool(d.Get("enable_advanced_threat_protection").(bool)),
```

here instead.

File: azurerm/resource_arm_storage_account.go [hunk 2]
```diff
@@ -548,6 +556,19 @@ func resourceArmStorageAccountCreate(d *schema.ResourceData, meta interface{}) e
...
+		},
+	}
+
+	_, err = advancedThreatProtectionClient.Create(ctx, d.Id(), advancedThreatProtectionSetting)
+	if err != nil {
```
Comment (katbyte) [inline azurerm/resource_arm_storage_account.go]: These lines should be merged:
```suggestion
	if _, err = advancedThreatProtectionClient.Create(ctx, d.Id(), advancedThreatProtectionSetting); err != nil {
```

File: azurerm/resource_arm_storage_account.go [hunk 3]
```diff
@@ -709,13 +732,30 @@ func resourceArmStorageAccountUpdate(d *schema.ResourceData, meta interface{}) e
 		d.SetPartial("network_rules")
 	}
 
+	if d.HasChange("enable_advanced_threat_protection") {
+		enableAdvanceThreatProtection := d.Get("enable_advanced_threat_protection").(bool)
+
+		opts := security.AdvancedThreatProtectionSetting{
+			AdvancedThreatProtectionProperties: &security.AdvancedThreatProtectionProperties{
+				IsEnabled: &enableAdvanceThreatProtection,
```
Comment (katbyte) [inline azurerm/resource_arm_storage_account.go]: As above:
```suggestion
				IsEnabled: &utils.Bool(d.Get("enable_advanced_threat_protection").(bool)),
```

File: azurerm/resource_arm_storage_account.go [hunk 4]
```diff
@@ -825,6 +865,14 @@ func resourceArmStorageAccountRead(d *schema.ResourceData, meta interface{}) err
 		return err
 	}
 
+	advancedThreatProtectionSetting, err := advancedThreatProtectionClient.Get(ctx, d.Id())
+
```
Comment (katbyte) [inline azurerm/resource_arm_storage_account.go]:
```suggestion
```

File: azurerm/resource_arm_storage_account.go [hunk 5]
```diff
@@ -825,6 +865,14 @@ func resourceArmStorageAccountRead(d *schema.ResourceData, meta interface{}) err
...
+	if err != nil {
+		return fmt.Errorf("Error reading the advanced threat protection settings of AzureRM Storage Account %q: %+v", name, err)
+	}
+
+	d.Set("enable_advanced_threat_protection", advancedThreatProtectionSetting.AdvancedThreatProtectionProperties.IsEnabled)
```
Comment (katbyte) [inline azurerm/resource_arm_storage_account.go]: We will need to check for nils here:
```suggestion
	if advancedThreatProtectionSetting != nil {
		if atpp := advancedThreatProtectionSetting.AdvancedThreatProtectionProperties; atpp != nil {
			d.Set("enable_advanced_threat_protection", atpp.IsEnabled)
		}
	} 
```

File: azurerm/resource_arm_storage_account_test.go [hunk 6]
```diff
@@ -713,6 +713,42 @@ func testCheckAzureRMStorageAccountDestroy(s *terraform.State) error {
...
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+			{
+				Config: postConfig,
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMStorageAccountExists(resourceName),
+					resource.TestCheckResourceAttr("azurerm_storage_account.testsa", "enable_advanced_threat_protection", "false"),
+				),
+			},
```
Comment (katbyte) [inline azurerm/resource_arm_storage_account_test.go]: Could we add another import step here at the end?

File: azurerm/resource_arm_storage_account.go (hunk 1 above)
Comment (benjamin37) [reply azurerm/resource_arm_storage_account.go]: done

File: azurerm/resource_arm_storage_account.go (hunk 2 above)
Comment (benjamin37) [reply azurerm/resource_arm_storage_account.go]: done

File: azurerm/resource_arm_storage_account.go (hunk 3 above)
Comment (benjamin37) [reply azurerm/resource_arm_storage_account.go]: done

File: azurerm/resource_arm_storage_account.go (hunk 4 above)
Comment (benjamin37) [reply azurerm/resource_arm_storage_account.go]: done

File: azurerm/resource_arm_storage_account.go (hunk 5 above)
Comment (benjamin37) [reply azurerm/resource_arm_storage_account.go]: done. Just one question about the first nil check. In all the other client calls the results are not checked for nil. E. g. `resp, err := client.GetProperties(ctx, resGroup, name, "")` in line `764`. Can you tell me why this check is required and the other one not?

File: azurerm/resource_arm_storage_account_test.go (hunk 6 above)
Comment (benjamin37) [reply azurerm/resource_arm_storage_account_test.go]: done

File: azurerm/resource_arm_storage_account.go (hunk 5 above)
Comment (benjamin37) [reply azurerm/resource_arm_storage_account.go]: Remove the first nil check again. `advancedThreatProtectionSetting` is not a pointer, so `&advancedThreatProtectionSetting` can't be nil. This also explains my question in the previous comment.

**Comment by benjamin37** [comment]
Hi @katbyte,

thanks for the fast reply and your feedback! I've commited the changes you requested. Just let me know if there is anything else to do for this pull request.

Kind regards

File: azurerm/resource_arm_storage_account.go (hunk 5 above)
Comment (katbyte) [reply azurerm/resource_arm_storage_account.go]: My bad, advancedThreatProtectionSetting not being resp threw me off, this is good now 🙂

**Pull request merged**